vllm/worker/multi_step_model_runner.py

```diff
@@ -0,0 +1,413 @@
+
+            assert seq_group.is_prompt is False  # No prompt
+            assert seq_group.prompt_logprob_indices == []  # No prompt
+            assert seq_group.sample_indices == [i]  # Simple
+            assert seq_group.seq_len is None  # Decode
+            assert seq_group.query_len is None  # Decode
+
+    def _advance_step(
+            self, model_input: ModelInputForGPUWithSamplingMetadata,
+            out: SamplerOutput) -> ModelInputForGPUWithSamplingMetadata:
+        # model_input.current_step += 1
+        assert model_input.seq_lens is not None
+        assert model_input.query_lens is not None
+        assert model_input.attn_metadata is not None
+
+        num_seqs = len(model_input.seq_lens)
+        num_queries = len(model_input.query_lens)
+        attn_metadata = model_input.attn_metadata
+        assert isinstance(attn_metadata, FlashAttentionMetadata)
+        self._update_flash_attn_metadata(attn_metadata, num_seqs, num_queries)
+
+        # Update GPU tensors
+        ops.advance_step(num_seqs=num_seqs,
+                         num_queries=num_queries,
+                         block_size=self.block_size,
+                         input_tokens=model_input.input_tokens,
+                         sampled_token_ids=out.sampled_token_ids,
+                         input_positions=model_input.input_positions,
+                         seq_lens=attn_metadata.seq_lens_tensor,
+                         slot_mapping=attn_metadata.slot_mapping,
+                         block_tables=attn_metadata.block_tables)
+
+        # Update sampling_metadata
+        # model_input.seq_lens = attn_metadata.seq_lens
+        # sampling_metadata = model_input.sampling_metadata
+        # self._update_sampling_metadata(sampling_metadata, num_seqs,
+        #                                num_queries)
+        for i in range(num_queries):
+            model_input.seq_lens[i] = attn_metadata.seq_lens[i]
+
+        return model_input
+
+
+def _pythonize_sampler_output(
+        model_input: ModelInputForGPUWithSamplingMetadata,
+        output: SamplerOutput) -> SamplerOutput:
+    # TODO(will): fix logprobs
+
+    # samples generation should have been skipped
+    assert not output.outputs
+
+    # print shapes
+    # print('output.sampled_token_ids', output.sampled_token_ids.shape)
+    # print('output.logprobs', output.logprobs.shape)
+    samples = torch.empty(*output.sampled_token_ids.shape,
+                          dtype=output.sampled_token_ids.dtype,
+                          device="cpu",
+                          pin_memory=True)
+    # prompt_logprobs = torch.empty(
+    #     *output.prompt_logprobs.shape,
+    #     dtype=output.prompt_logprobs.dtype,
+    #     device="cpu",
+    #     pin_memory=True)
+
+    # CPU GPU sync
+    # logprobs = logprobs.copy_(output.logprobs, non_blocking=False)
+    samples = samples.copy_(output.sampled_token_ids, non_blocking=False)
+
+    samples_list = samples.tolist()
```

Comment:
I think tolist() is blocking, no?

Reply:
Yes, at some point we need to incur this blocking call but ideally this will happen when the CPU is ahead of GPU. Will add comments explaining this